photon-lib/src/main/scala/com/linkedin/photon/ml/function/L2Regularization.scala

```diff
@@ -59,16 +67,22 @@ trait L2Regularization extends ObjectiveFunction {
       input: Data,
       coefficients: Coefficients,
       normalizationContext: BroadcastWrapper[NormalizationContext]): Double =
-    super.value(input, coefficients, normalizationContext) + l2RegValue(convertToVector(coefficients))
+    super.value(input, coefficients, normalizationContext) + l2RegValue(convertToVector(coefficients), interceptIndex)
 
   /**
    * Compute the L2 regularization value for the given model coefficients.
    *
    * @param coefficients The model coefficients
    * @return The L2 regularization value
    */
-  protected def l2RegValue(coefficients: Vector[Double]): Double =
-    l2RegWeight * coefficients.dot(coefficients) / 2
+  protected def l2RegValue(coefficients: Vector[Double], interceptIndex: Int = -1): Double = {
+    val l2Regularization = interceptIndex match {
+      case -1 => l2RegWeight * coefficients.dot(coefficients) / 2
+      case x if x >= 0 => l2RegWeight * (coefficients.dot(coefficients) - pow(coefficients(x), 2)) / 2
+      case other => throw new IllegalArgumentException(s"Encountered invalid intercept index $other")
```

Comment:
We can remove this check by having a `require` statement above the L2 weight getter:
```
require(interceptIndexOpt.forall(_ >= 0), "Invalid intercept index ${interceptIndexOpt.get}")
```

Reply:
This link explains the reason for calling `forall`:
https://blog.tmorris.net/posts/scalaoption-cheat-sheet/

Reply:
Is it always true that `INTERCEPT_KEY` exists in `featureIndexMapLoaders`? I set "-1" as default value when it does not exists, so no subtraction. 

Besides, I may need to add another condition that `_ < coefficients.size` and throw exceptions accordingly.

Reply:
No, an intercept is not guaranteed to exist in `featureIndexMapLoaders`. So the code to fetch it will need to look something like this (in the `GameTrainingDriver`):
```
val featureShardIndices = featureIndexMapLoaders.mapValues { indexMap =>
  indexMap.indexMapForDriver().getIndex(Constants.INTERCEPT_KEY) match {
    case IndexMap.NULL_KEY => None
    case i if i >= 0 => Some(i)
    case i => throw new IndexOutOfBoundsException("...")
  }
}
val coordinateInterceptIndices = getRequiredParam(coordinateConfigurations).mapValues { config =>
  featureShardIndices(config.dataConfiguration.featureShardId)
}
```

You're right - we should do a safety check against `coefficients.size`. Actually, I just realized that because of random effect projection, the index of the intercept will change for each random effect. Let's take a look together later, I'll show you what I mean.